File: C-plus-plus/Stack/Prefix-Postfix/stack.cpp
#include "node.cpp"
#include <iostream>

using namespace std;

class Stack {
public:
    Node* top;

    Stack() {
        top = nullptr;
    }

    bool isEmpty() {
        return top == nullptr;
    }

    Operator* peek() {
        return top->op;
    }

    void push(Node *node) {
        if (isEmpty()) {
            top = node;
        }
        else {
            node->down = top;
            top = node;
        }
    }

    void pop() {
        if (isEmpty()) {
            cout << "Stack is empty" << endl;
        }
        else {
            Node* deleted = top;
            top = top->down;
            delete deleted;
        }
    }


    void show() {
        Node* tmp = top;

        while (tmp != nullptr) {
            cout << tmp->value << endl;
            tmp = tmp->down;
        }
    }
};

File: Python/PostfixExpression/stack.py
from node import Node

class Stack:


    def __init__(self):
        self.stack = list()
        self.top:Node = None


    def isEmpty(self):
        return self.top is None


    def peek(self):
        return self.top


    def push(self, node:Node):
        if self.isEmpty():
            self.top = node
        else:
            node.down = self.top
            self.top = node
        print("Node added")


    def pop(self):
        if self.isEmpty():
            print("Stack is empty !")
        else:
            deleted = self.top
            self.top = self.top.down
            return deleted
        print("Node deleted")